File: neet_code_all_cpp/math_and_geometry/shift_2d_grid.cpp
#include <iostream>
#include <vector>
#include <memory>

class Solution {
    public:
        // AC
        std::vector<std::vector<int>> shiftGrid(std::vector<std::vector<int>> grid, int k) {
            auto [m, n] = std::pair(grid.size(), grid[0].size());
            auto prev = grid; 
            auto result = prev;

            for(int iter = 0; iter < k ; ++iter) {
                for(int i = 0; i < m; ++i) {
                    for(int j = 0; j < n; ++j) {
                        if(j + 1 < n) {
                            result[i][j + 1] = prev[i][j];
                        }
                    }
                }
                for(int i = 0; i < m; ++i) {
                    if(i + 1 < m) {
                        result[i + 1][0] = prev[i][n - 1];
                    }
                }
                result[0][0] = prev[m - 1][n - 1];
                prev = result;
            }

            return result;
        }
};

// 模範解答
class SolutionAns {
    public:
        std::vector<std::vector<int>> shiftGrid(std::vector<std::vector<int>>& grid, int k) {
            int m = grid.size(), n = grid[0].size();
            auto posToVal = [&](int r, int c) {
                return r * n + c;
            };
            // 生のポインタをnewしてもstd::vector<T>でもunique_ptr<T[]>でもok
            // 速度は 生 > unique_ptr<T []> > std::vector<T> 
            auto valToPos = [&](int v) {
                //return std::vector{v/n, v % n};
                //return new int[]{v/n, v % n};
                auto ptr = std::make_unique<int[]>(2);
                ptr[0] = v / n;
                ptr[1] = v % n;

                return ptr;
            };

            std::vector<std::vector<int>> result;
            for(int r = 0; r < m; ++r) {
                std::vector<int> row;
                for(int c = 0; c < n; ++c) {
                    row.push_back(0);
                }
                result.push_back(row);
            }
            for(int r = 0; r < m; ++r) {
                for(int c= 0; c < n; ++c) {
                    int newVal = (posToVal(r, c) + k) % (m * n);
                    auto newRc = valToPos(newVal);
                    result[newRc[0]][newRc[1]] = grid[r][c];
                }
            }

            return result;
        }

        // AC
        std::vector<std::vector<int>> shiftGrid2(std::vector<std::vector<int>>& grid, int k) {
            int m = grid.size(), n = grid[0].size();
            auto posToVal = [&](int r, int c) {
                return r * n + c;
            };
            // std::pair<U,V>を返す
            auto valToPos = [&](int v) {
                return std::pair(v / n,  v % n);
            };

            std::vector<std::vector<int>> result;
            for(int r = 0; r < m; ++r) {
                std::vector<int> row;
                for(int c = 0; c < n; ++c) {
                    row.push_back(0);
                }
                result.push_back(row);
            }
            for(int r = 0; r < m; ++r) {
                for(int c= 0; c < n; ++c) {
                    int newVal = (posToVal(r, c) + k) % (m * n);
                    auto [idx_r, idx_c] = valToPos(newVal);
                    result[idx_r][idx_c] = grid[r][c];
                }
            }

            return result;
        }
};

void show_result(const std::vector<std::vector<int>>& result) {
    for(auto &v: result) {
        for(auto &w: v) {
            std::cout << w << " ";
        }
        std::cout << '\n';
    }
}

int main(void) {
    using Case = std::pair<std::vector<std::vector<int>>, int>;

    Case case_1 = {{{1,2,3},{4,5,6},{7,8,9}}, 1};
    Case case_2 = {{{3,8,1,9},{19,7,2,5},{4,6,11,10},{12,0,21,13}}, 4};
    Case case_3 = {{{1,2,3},{4,5,6},{7,8,9}}, 9};

    Solution s_1;
    auto res_1 = s_1.shiftGrid(case_1.first, case_1.second);
    auto res_2 = s_1.shiftGrid(case_2.first, case_2.second);
    auto res_3 = s_1.shiftGrid(case_3.first, case_3.second);
    show_result(res_1);
    show_result(res_2);
    show_result(res_3);

    SolutionAns s_ans;
    auto res_1_ans = s_ans.shiftGrid(case_1.first, case_1.second);
    auto res_2_ans = s_ans.shiftGrid(case_2.first, case_2.second);
    auto res_3_ans = s_ans.shiftGrid(case_3.first, case_3.second);
    show_result(res_1_ans);
    show_result(res_2_ans);
    show_result(res_3_ans);
}

File: neet_code_all/src/math_and_geometry/shift_2d_grid.rs
struct Solution;
impl Solution {
    // AC
    pub fn shift_grid(grid: Vec<Vec<i32>>, k: i32) -> Vec<Vec<i32>> {
        let (m, n) = (grid.len(), grid[0].len());
        let mut prev = grid;
        let mut result = prev.clone();

        for _iter in 0..k {
            for i in 0..m {
                for j in 0..n {
                    if j + 1 < n {
                        result[i][j + 1] = prev[i][j];
                    }
                }
            }
            for i in 0..m {
                if i + 1 < m {
                    result[i + 1][0] = prev[i][n - 1];
                }
            }
            result[0][0] = prev[m - 1][n - 1];

            prev = result.clone();
        }

        result
    }
}

// C++の模範解答より
struct SolutionAnsCpp;
impl SolutionAnsCpp {
    // AC
    pub fn shift_grid(grid: Vec<Vec<i32>>, k: i32) -> Vec<Vec<i32>> {
        let (m, n) = (grid.len(), grid[0].len());
        let pos_to_val = |r, c| r * n + c;
        let val_to_pos = |v| (v / n, v % n);

        let mut result = vec![];
        for _r in 0..m {
            let mut row = vec![];
            for _c in 0..n {
                row.push(0);
            }
            result.push(row);
        }
        for r in 0..m {
            for c in 0..n {
                let new_val = (pos_to_val(r, c) + k as usize) % (m * n);
                let (idx_r, idx_c) = val_to_pos(new_val);
                result[idx_r][idx_c] = grid[r][c];
            }
        }

        result
    }
}

fn main() {
    let case_1 = (vec![vec![1, 2, 3], vec![4, 5, 6], vec![7, 8, 9]], 1);
    // => [[9,1,2],[3,4,5],[6,7,8]]
    let case_2 = (
        vec![
            vec![3, 8, 1, 9],
            vec![19, 7, 2, 5],
            vec![4, 6, 11, 10],
            vec![12, 0, 21, 13],
        ],
        4,
    );
    // => [[12,0,21,13],[3,8,1,9],[19,7,2,5],[4,6,11,10]]
    let case_3 = (vec![vec![1, 2, 3], vec![4, 5, 6], vec![7, 8, 9]], 9);
    // => [[1,2,3],[4,5,6],[7,8,9]]

    println!(
        "case_1: {:?}",
        Solution::shift_grid(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {:?}",
        Solution::shift_grid(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {:?}",
        Solution::shift_grid(case_3.0.clone(), case_3.1)
    );

    println!(
        "case_1: {:?}",
        SolutionAnsCpp::shift_grid(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {:?}",
        SolutionAnsCpp::shift_grid(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {:?}",
        SolutionAnsCpp::shift_grid(case_3.0.clone(), case_3.1)
    );
}
